rules/noUnnecessaryElseRule.ts

```diff
@@ -1,22 +1,56 @@
-import * as ts from 'typescript';
 import * as Lint from 'tslint';
 import * as utils from 'tsutils';
-import { AbstractIfStatementWalker } from '../src/walker';
+import * as ts from 'typescript';
 import { isElseIf } from '../src/utils';
 
 const FAIL_MESSAGE = `unnecessary else`;
+const FAIL_MESSAGE_BLOCK = `unnecessary else block`;
+
+const elseBraceRegex = /^[ \t]*else[ \t]\{(.*)\}[ \t]*$/;
+const elseRegex = /^[ \t]*else(.*)$/;
 
 export class Rule extends Lint.Rules.AbstractRule {
-    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
-        return this.applyWithWalker(new IfWalker(sourceFile, this.ruleName, undefined));
-    }
+  public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+    return this.applyWithWalker(new IfWalker(sourceFile, this.ruleName, undefined));
+  }
 }
 
-class IfWalker extends AbstractIfStatementWalker<void> {
-    protected _checkIfStatement(node: ts.IfStatement) {
-        if (node.elseStatement !== undefined &&
-            !isElseIf(node) &&
-            utils.endsControlFlow(node.thenStatement))
-            this.addFailureAtNode(node.getChildAt(5 /*else*/, this.sourceFile), FAIL_MESSAGE);
+class IfWalker extends Lint.AbstractWalker<void> {
+  public walk(sourceFile: ts.SourceFile) {
+    const cb = (node: ts.Node): void => {
+      if (node.kind === ts.SyntaxKind.IfStatement && !this._checkIfStatement(<ts.IfStatement>node))
+        return;
+      return ts.forEachChild(node, cb);
+    };
+
+    return ts.forEachChild(sourceFile, cb);
+  }
+
+  /** Return true to traverse children, false otherwise */
+  private _checkIfStatement(node: ts.IfStatement): boolean {
+    const elseStatement = node.elseStatement;
+
+    if (isElseStatement(elseStatement) && !isElseIf(node) && utils.endsControlFlow(node.thenStatement)) {
+      // remove else scope if safe: keep blocks where local variables change scope when unwrapped
+      if (isElseStatement(elseStatement) && utils.isBlock(elseStatement) && !hasLocals(elseStatement)) {
+        const replacement = elseStatement.getText().replace(elseBraceRegex, '$1').replace(elseRegex, '$1');
+        const fixBlock = Lint.Replacement.replaceNode(elseStatement, replacement);
+        this.addFailureAtNode(node.getChildAt(5 /*else*/, this.sourceFile), FAIL_MESSAGE_BLOCK, fixBlock);
+        return true; // remove the block scope buggers with offsets in children, don't process them now.
+      }
+
+      // remove else
+      const fixElse = Lint.Replacement.replaceFromTo(node.thenStatement.getEnd(), elseStatement.getStart(), '');
```

Comment:
Consider just removing the `else` keyword and leave all the whitespace unchanged.

```ts
if (a)
  a()
else
  b()

// should be fixed to
if (a)
  a()
  b()

// instead of producing invalid code
if (a)
  a()b()
```